_posts/2025-04-19-navigating-pagination-with-hybrid-query.md

```diff
@@ -0,0 +1,229 @@
+
+Traditional queries (like match and term) use a simple `from` + `size` formula to determine the maximum number of search results to retrieve from each shard. However, applying this same formula to hybrid search creates challenges that affect result accuracy, or _ground truth_. Ground truth refers to information from direct observation rather than inference; in practice, it represents the best available test or benchmark under reasonable conditions. This serves as a foundation for validating data accuracy, whether through customer feedback, human labels, or the most reliable available testing methods even when classification may be imperfect.
+
+Let's examine an example. Consider a hybrid query search request with these parameters:
+- `from = 0` 
+- `size = 3`
+
+The request contains two subqueries under the hybrid clause: a match query and a k-NN query. The normalization processor is configured to assign a greater weight to the match query (0.7) higher than the k-NN query (0.3), as shown in the following diagram.
+
+![higher-weightage](/assets/media/blog-images/2025-04-19-navigating-pagination-with-hybrid-query/higher-weightage.png){:class="img-centered"}
+
+### How the from + size formula affects ground truth
+
+Modifying `from` or ``size`` values increases the number of results retrieved for each subquery from a shard. However, these new results don't simply append to the last page. They can appear earlier in the ranked list, changing the ground truth you're trying to paginate through. This happens in the following two main scenarios:
+
+1. **Higher-weighted subquery effect**:
+    - New results from a higher-weighted subquery may rank higher after score combination.
+    - As a result, these new entries may appear earlier in the final result set---either on the first few pages or somewhere in the middle---rather than at the end, disrupting the expected pagination order.
+
+    In our example, increasing size to 4 causes an unexpected result: the new Query 1 result (document ID 14) appears as the fourth document instead of at the end, because of Query 1's higher weight during normalization, as shown in the following diagram.
+
+    ![more-weightage](/assets/media/blog-images/2025-04-19-navigating-pagination-with-hybrid-query/more-weightage.png){:class="img-centered"}
+
+1. **Duplicate result effect**:
+    - Newly retrieved results can become duplicates, meaning they exist in both subquery results.
+    - After the score combination process, the combined score of a duplicate entry increases, making it more relevant.
+    - With equal subquery weights, duplicate results rank higher than expected, appearing earlier in search results.
+
+    In the following diagram, documents with IDs 0 and 12 are duplicate results appearing in both match and k-NN subqueries. Because duplicate documents typically receive higher hybrid scores than single-occurrence documents, they appear higher in the results, disrupting the expected pagination order.
```

Comment:
```suggestion
    In the following diagram, the documents with the IDs 0 and 12 are duplicate results appearing in both the match and k-NN subqueries. Because duplicate documents typically receive higher hybrid scores than unduplicated documents, they appear higher in the results, disrupting the expected pagination order.
```